proptest-state-machine/src/strategy.rs

```diff
@@ -0,0 +1,761 @@
+}
+
+impl<
+        State: Clone + Debug,
+        Transition: Clone + Debug,
+        StateValueTree: ValueTree<Value = State>,
+        TransitionValueTree: ValueTree<Value = Transition>,
+    >
+    SequentialValueTree<State, Transition, StateValueTree, TransitionValueTree>
+{
+    /// Try to apply the next `self.shrink`. Returns `true` if a shrink has been
+    /// applied.
+    fn try_simplify(&mut self) -> bool {
+        if let DeleteTransition(ix) = self.shrink {
+            if self.included_transitions.count() == self.min_size {
+                // Can't delete any more transitions, move on to shrinking them
+                self.shrink = Transition(0);
+            } else {
+                // Delete the index from the included transitions
+                self.included_transitions.clear(ix);
+
+                self.last_shrink = Some(self.shrink);
+                self.shrink = if ix == 0 {
+                    // Reached the beginning of the list, move on to shrinking
+                    Transition(0)
+                } else {
+                    // Try to delete the previous transition next
+                    DeleteTransition(ix - 1)
+                };
+                // If this delete is not acceptable, undo it and try again
+                if !self.check_acceptable(None) {
+                    self.included_transitions.set(ix);
+                    self.last_shrink = None;
+                    return self.try_simplify();
+                }
+                // If the delete was accepted, remove this index from shrinkable
+                // transitions
+                self.shrinkable_transitions.clear(ix);
+                return true;
+            }
+        }
+
+        while let Transition(ix) = self.shrink {
+            if self.shrinkable_transitions.count() == 0 {
+                // Move on to shrinking the initial state
+                self.shrink = Shrink::InitialState;
+                break;
+            }
+
+            if !self.included_transitions.test(ix) {
+                // No use shrinking something we're not including
+                self.shrink = self.next_shrink_transition(ix);
+                continue;
+            }
+
+            if let Some((SimplifyRejected, _trans)) =
+                self.acceptable_transitions.get(ix)
+            {
+                // This transition is already simplified and rejected
+                self.shrink = self.next_shrink_transition(ix);
+            } else if self.transitions[ix].simplify() {
+                self.last_shrink = Some(self.shrink);
+                if self.check_acceptable(Some(ix)) {
+                    self.acceptable_transitions[ix] =
+                        (Current, self.transitions[ix].current());
+                    return true;
+                } else {
+                    let (state, _trans) =
+                        self.acceptable_transitions.get_mut(ix).unwrap();
+                    *state = SimplifyRejected;
```

Comment:
Why is it that we stop simplifying a transition as soon as we have a simplification that fails `check_acceptable`, should we not continue simplifying until we have an acceptable simplification which will later be `complicated()`?

The example from https://proptest-rs.github.io/proptest/proptest/tutorial/shrinking-basics.html continues to simplify on `// Still failing, find a simpler case` until the test succeeds or can no longer be simplified further.

Reply:
Reading this
```
/// Hence, the `preconditions` only needs to contain checks for invariants that
/// depend on the current state and may be broken by shrinking and it doesn't
/// need to cover invariants that do not depend on the current state.
```
adds some clarity, I suppose if the test can only ever operate within the transition graph described by the preconditions then there is no simplification allowed which breaks preconditions too

Reply:
Thinking through this it's still unclear to me. If a test fails because of an invalid transition given a state then wouldn't we'd want to simplify that transition as much as possible?

Reply:
If a shrink of a transition doesn't meet preconditions then we have to *omit* that change from the test as it's not a valid input. Sometimes a transition might become acceptable after a transition that comes before it in the sequence gets shrunk.

Together with the rotation on line 470, I experimentally found that this seems to shrink the transitions better than other strategies I tried before. The rotation ensures that we go back to transitions for which a shrink might have been previously rejected and it will attempt to simplify transitions as much as possible. I hope that answers those questions?

Reply:
I think I get what you're saying. Transitions form a causal chain of events, preconditions represent what paths of connected transitions can exist in that chain, and since causal chains are DAGs, our current set of transitions is one path in that DAG over the set of all _possible_ transitions (those which satisfy preconditions). Simplifying a transition may cause the test to enter a different path in the DAG overall so we have to do additional passes over the transition chain to test that it still meets preconditions.

I guess one thing that's still not entirely clear to me, are not failing preconditions a part of the test? If I have transitions A -> B -> C, and `B` fails preconditions from the start, then wouldn't I be encouraged to simplify `B` as much as possible until it succeeds and then complicate it (or just return it to its last failing state)? `B` is essentially a mediator of C, so observations on B screens off observations of C all else being equal, so we don't need to care about C at that point. I think what you may be saying is we simplify `A` first, and if that fails preconditions then we're screening off `B`, but `B` was what was failing the test to begin with and therefore we're no longer going to capture the original cause of test failure. Assuming I'm getting that right I think that makes sense, and possibly is a side-effect of the process we're using here to capture failure. Something we could possibly revisit in the future. 

(FYI, the reason I consider this a DAG is because i'm modeling this in my head via causality. E.g. Given a set of transitions InitialState -> A -> B -> D -> A (again), the causal notion should be InitialState -> A(t1) -> B(t1) -> D -> A(t2) -> B(t2) where observing the failure at B(t1) factors out everything that follows [inspired by [this](https://en.wikipedia.org/wiki/Causal_model)]).

Reply:
That's an interesting way to think about it! I think your causality model is on the right way, but any failing preconditions are in fact not part of the test (it is an invalid input for the test) and the initial transition sequence must always pass preconditions (this is ensured in `Sequential`'s `Strategy::new_tree` impl via filtering). What the shrinking process does is when it breaks a precondition is:

- for `Shrink::DeleteTransition` undoes deletion
- for `Shrink::Transition(_)` or `InitialState`  it ignores the simplification or complication and instead continues to use the state of the transition or of the initial state as it was before the shrink

Breaking a precondition on `Shrink::Transition` however doesn't stop the shrinking process and it will try to use that shrunk transition again next time the sequence rotation gets to it. In the example of `A -> B -> C`, if `B` shrink fails a precondition, `B` stays unchanged for that round. The shrinking process may get to it again after rotation and after `A` is shrunk it may (or may not) "unblock" the previously shrunk `B`'s strategy to become acceptable.